### PR TITLE
kubeadm: add a preflight check for Docker and cgroup driver

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:
systemd is the recommended driver as per the setup of running
the kubelet using systemd as the init system. Add a preflight
check that throws a warning if this isn't the case.

**Which issue(s) this PR fixes**:

xref kubernetes/kubeadm#1394

**Special notes for your reviewer**:
NONE

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: add a preflight check that throws a warning if the cgroup driver for Docker on Linux is not "systemd" as per the k8s.io CRI installation guide.
```

/assign @timothysc 
/cc @rosti 
cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/kind bug
/priority critical-urgent

can be of `kind feature` too.

### CODE REVIEW

**Comment by neolit123** [comment]
example output:
```
[init] Using Kubernetes version: v1.13.3
[preflight] Running pre-flight checks
	[WARNING IsDockerSystemdCheck]: detected "cgroupfs" as the Docker cgroup driver. The recommended driver is "systemd". Please follow the guide at https://kubernetes.io/docs/setup/cri/
```

**Review by yagonobre — APPROVED** [review]
Thanks @neolit123 
/lgtm
/hold
feel free to unhold

**Comment by neolit123** [comment]
let's have @timothysc and @rosti have a look too.

File: cmd/kubeadm/app/preflight/checks_darwin.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// +build darwin
```
Comment (timothysc) [inline cmd/kubeadm/app/preflight/checks_darwin.go]: Might want to add a comment about mac below, my brain kind of SEGV'd when reading this at first.

File: cmd/kubeadm/app/preflight/checks_linux.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import (
+	"github.com/pkg/errors"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util"
+	"k8s.io/utils/exec"
+)
+
+// Check validates if Docker is setup to use systemd as the cgroup driver.
+func (idsc IsDockerSystemdCheck) Check() (warnings, errorList []error) {
+	warnings = []error{}
+	driver, err := util.GetCgroupDriverDocker(exec.New())
```
Comment (timothysc) [inline cmd/kubeadm/app/preflight/checks_linux.go]: what about other CRIs, or is that abstracted by the utility function?

**Review by timothysc — COMMENTED** [review]
minor comments/nits otherwise lgtm 

@mauilion - is this helpful? Or would you like to see more info?

File: cmd/kubeadm/app/preflight/checks_linux.go (hunk 2 above)
Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks_linux.go]: we cannot detect cgroup driver for other CRIs

File: cmd/kubeadm/app/preflight/checks_darwin.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: do you mean adding a note that `darwin == macos` in golang?

File: cmd/kubeadm/app/preflight/checks.go [hunk 3]
```diff
@@ -998,6 +1006,9 @@ func addCommonChecks(execer utilsexec.Interface, cfg kubeadmapi.CommonConfigurat
 		if containerRuntime.IsDocker() {
 			isDocker = true
 			checks = append(checks, ServiceCheck{Service: "docker", CheckIfActive: true})
+			if runtime.GOOS == "linux" {
```
Comment (detiber) [inline cmd/kubeadm/app/preflight/checks.go]: Why both the runtime and buildtime conditionals for the check?

File: cmd/kubeadm/app/preflight/checks_darwin.go (hunk 1 above)
Comment (detiber) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: Do we actually need the conditional here for a mac? Do we expect users to run `kubeadm` on a mac?

File: cmd/kubeadm/app/preflight/checks_linux.go (hunk 2 above)
Comment (timothysc) [reply cmd/kubeadm/app/preflight/checks_linux.go]: Could you add a //TODO here then and open an issue for that.

File: cmd/kubeadm/app/preflight/checks_darwin.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: we have a typecheck from hack that will fail if the symbol is not defined for darwin.

File: cmd/kubeadm/app/preflight/checks.go (hunk 3 above)
Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks.go]: sure, we can call this as it's a no-op for darwin and windows.

File: cmd/kubeadm/app/preflight/checks_linux.go (hunk 2 above)
Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks_linux.go]: issue is here:
https://github.com/kubernetes/kubeadm/issues/874

will add TODO

File: cmd/kubeadm/app/preflight/checks_darwin.go (hunk 1 above)
Comment (timothysc) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: I think that's the #1 development platform, so we should probably do right by the developers.

Comment (timothysc) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: @neolit123 - just a comment below is fine. 
// This is a macos stub, yadda yadda.

**Comment by neolit123** [comment]
@timothysc @detiber updated

**Pull request merged**